shared-bindings/wdt/WDT.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016 Paul Sokolovsky
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "py/obj.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/wdt/__init__.h"
+#include "shared-bindings/wdt/WDT.h"
+
+static wdt_wdt_obj_t *wdt_singleton;
+
+//| class WDT:
+//|     """Watchdog Timer"""
+//|
+//|     def __init__(self, ):
+//|         """This class represents the system's Watchdog Timer. It is a
+//|         singleton and will always return the same instance."""
+//|         ...
+//|
+STATIC mp_obj_t wdt_wdt_make_new(const mp_obj_type_t *type, size_t n_args,
+                                 const mp_obj_t *pos_args,
+                                 mp_map_t *kw_args) {
+    enum { ARG_timeout_ms, ARG_sleep };
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_timeout_ms, MP_ARG_INT | MP_ARG_REQUIRED},
+        {MP_QSTR_sleep, MP_ARG_BOOL, {.u_bool = false}},
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+
+    if (wdt_singleton)
+        return wdt_singleton;
```

Comment:
Is it possible to change the time and sleep settings? It'd be weird to pass them in to no effect. Maybe this should be a singleton on `supervisor` or `microcontroller` instead similar to `microcontroller.cpu`.

Reply:
I can place it there, but does that mean that we need to have a Watchdog on every platform? Or will it be like NVM, and will default to `mp_const_none_obj` if `CIRCUITPY_WATCHDOG` is `0`?

On NRF, it is not possible to disable the WDT, nor is it possible to adjust the settings -- the WDT registers are only unlocked after a reset.

Reply:
I was thinking it'd be like nvm where it is None when CIRCUITPY_WATCHDOG = 0.